extensions-core/parquet-extensions/src/main/java/org/apache/druid/data/input/parquet/DruidParquetReader.java

```diff
@@ -0,0 +1,182 @@
+  private final ObjectFlattener<Group> flattener;
+  private final byte[] buffer = new byte[InputEntity.DEFAULT_FETCH_BUFFER_SIZE];
+
+  private final ParquetGroupConverter converter;
+  private final ParquetGroupJsonProvider jsonProvider;
+
+  private final ParquetReader<Group> reader;
+  private final ParquetMetadata metadata;
+  private final Closer closer;
+
+  public DruidParquetReader(
+      InputRowSchema inputRowSchema,
+      InputEntity source,
+      File temporaryDirectory,
+      JSONPathSpec flattenSpec,
+      boolean binaryAsString
+  ) throws IOException
+  {
+    this.inputRowSchema = inputRowSchema;
+    this.converter = new ParquetGroupConverter(binaryAsString);
+    this.jsonProvider = new ParquetGroupJsonProvider(converter);
+    this.flattener = ObjectFlatteners.create(flattenSpec, new ParquetGroupFlattenerMaker(binaryAsString));
+
+    closer = Closer.create();
+    final InputEntity.CleanableFile file = closer.register(source.fetch(temporaryDirectory, buffer));
+    final Path path = new Path(file.file().toURI());
+
+    final ClassLoader currentClassLoader = Thread.currentThread().getContextClassLoader();
+    try {
+      Thread.currentThread().setContextClassLoader(getClass().getClassLoader());
+      reader = closer.register(ParquetReader.builder(new GroupReadSupport(), path).build());
+      metadata = ParquetFileReader.readFooter(new Configuration(), path);
+    }
+    finally {
+      Thread.currentThread().setContextClassLoader(currentClassLoader);
+    }
+  }
+
+  @Override
+  protected CloseableIterator<Group> intermediateRowIterator() throws IOException
+  {
+    return new CloseableIterator<Group>()
+    {
+      Group value = null;
+
+      @Override
+      public boolean hasNext()
+      {
+        if (value == null) {
+          try {
+            value = reader.read();
+          }
+          catch (IOException e) {
+            throw new RuntimeException(e);
+          }
+        }
+        return value != null;
+      }
+
+      @Override
+      public Group next()
+      {
+        if (value == null) {
+          throw new NoSuchElementException();
+        }
+        Group currentValue = value;
+        value = null;
+        return currentValue;
+      }
+
+      @Override
+      public void close() throws IOException
+      {
+        closer.close();
+      }
+    };
+  }
+
+  @Override
+  protected List<InputRow> parseInputRows(Group intermediateRow) throws IOException, ParseException
+  {
+    return Collections.singletonList(
+        MapInputRowParser.parse(
+            inputRowSchema.getTimestampSpec(),
+            inputRowSchema.getDimensionsSpec(),
+            flattener.flatten(intermediateRow)
+        )
+    );
+  }
+
+  @Override
+  protected String toJson(Group intermediateRow) throws IOException
+  {
+    Object converted = convertObject(intermediateRow);
+    return DEFAULT_JSON_WRITER.writeValueAsString(converted);
+  }
+
+  private Object convertObject(Object o)
```

Comment:
`normalizeObjectForJson`?

Reply:
Actually, this is a pretty cool method that seems like it'd be useful for Avro / ORC too. Could you consider putting it in core so all the extensions can use it?